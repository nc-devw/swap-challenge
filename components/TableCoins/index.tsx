import { AxiosResponse } from "axios";
import React from "react";
import { useAppContext } from "../../context/global";
import useRestClient from "../../hooks/useRestClient";
import { Coin } from "../../services/transforms/coins";
import Row from "./row";

type Props = {};

export interface CoinsBalance extends Coin {
  currentQuantity: number;
  currentUsdPrice?: number;
}

const TableCoins = (props: Props) => {
  const { response }: { response: AxiosResponse<Coin[], any> | undefined } =
    useRestClient({
      method: "GET",
      url: `/coins`,
    });

  const { user } = useAppContext();

  const [balance, setBalance] = React.useState(0);
  const [coins, setCoins] = React.useState<CoinsBalance[]>([]);

  React.useEffect(() => {
    if (
      response?.data.length === 0 ||
      user?.coins.length === 0 ||
      coins.length !== 0
    ) {
      return;
    }

    response?.data.forEach((coinData) => {
      const coin = user?.coins?.find((coin) => coin.symbol === coinData.symbol);

      if (!coin) {
        let newArray = [...coins, { ...coinData, currentQuantity: 0 }];
        setCoins(newArray);
        return;
      }
      const currentUsdPrice = coin.quantity * coinData.usdPrice;
      setBalance((prevState) => prevState + currentUsdPrice);

      setCoins((prevState) => [
        ...prevState,
        {
          ...coinData,
          currentQuantity: coin.quantity,
          currentUsdPrice,
        },
      ]);

      return;
    });

    return;
  }, [response, user]);
  return (
    <div className="w-100 rounded-lg m-12 shadow">
      <div className="flex flex-col">
        <div className="text-center p-4 shadow">
          <p className="uppercase font-bold text-lg">Your assets</p>
          <p className="uppercase text-sm text-gray-400">{`Total Balance: ${balance} USD`}</p>
        </div>
        {coins.length > 0 &&
          coins.map((coin: CoinsBalance) => (
            <Row key={coin.symbol} {...coin} />
          ))}
      </div>
    </div>
  );
};

export default TableCoins;
