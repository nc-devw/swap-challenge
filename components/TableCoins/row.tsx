import React from "react";
import Image from "next/image";
import { Coin } from "../../context/global";

type RowProps = {
  iconUrl: string;
  name: string;
  symbol: string;
  usdPrice: number;
  userCoins?: Coin[];
};

interface current {
  currentQuantity: number;
  currentUsdPrice?: number;
}

const Row = ({ iconUrl, name, symbol, usdPrice, userCoins }: RowProps) => {
  const convertCoinToUsd: () => current = React.useCallback(() => {
    const coin = userCoins?.find((coin) => coin.symbol === symbol);

    if (!coin) {
      return { currentQuantity: 0 };
    }

    return {
      currentQuantity: coin.quantity,
      currentUsdPrice: coin.quantity * usdPrice,
    };
  }, [symbol, usdPrice, userCoins]);

  const { currentQuantity, currentUsdPrice } = convertCoinToUsd();

  return (
    <div className="flex justify-between p-6 shadow">
      <div className="flex items-center">
        <div className="h-9 w-9">
          <Image src={iconUrl} alt={symbol} width="100%" height="100%" />
        </div>
        <div className="ml-2 flex flex-col justify-center text-sm">
          <div className="font-bold">{name}</div>
          <div className="uppercase">{symbol}</div>
        </div>
      </div>
      <div className="text-right">
        <div className="font-bold">$ {currentUsdPrice}</div>
        <div className="font-light text-sm text-gray-400">{`${currentQuantity} ${symbol}`}</div>
      </div>
    </div>
  );
};

export default Row;
