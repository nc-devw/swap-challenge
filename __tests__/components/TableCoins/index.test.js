import { render } from "@testing-library/react";
import TableCoins from "@/components/TableCoins";

const mockCoins = [
  {
    name: "Bitcoin",
    symbol: "btc",
    iconUrl:
      "https://assets.coingecko.com/coins/images/1/small/bitcoin.png?1547033579",
    usdPrice: 31872,
  },
  {
    name: "USD Coin",
    symbol: "usdc",
    iconUrl:
      "https://assets.coingecko.com/coins/images/6319/small/USD_Coin_icon.png?1547042389",
    usdPrice: 1,
  },
];

describe("Table Coins", () => {
  it("should render correctly", () => {
    const { asFragment } = render(<TableCoins coins={mockCoins} />);
    expect(asFragment).toMatchSnapshot();
  });
});
