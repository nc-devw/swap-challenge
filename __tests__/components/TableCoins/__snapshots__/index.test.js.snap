// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Coins should render correctly 1`] = `
<DocumentFragment>
  <div
    class="rounded-lg m-12 border shadow bg-white undefined"
  >
    <div
      class="flex flex-col"
    >
      <div
        class="text-center p-4 shadow"
      >
        <p
          class="uppercase font-bold text-lg"
        >
          Your assets
        </p>
        <p
          class="uppercase text-sm text-gray-400"
        >
          Total Balance: 63.744 USD
        </p>
      </div>
      <div
        class="flex justify-between p-6 shadow"
      >
        <div
          class="flex items-center"
        >
          <div
            class="h-9 w-9"
          >
            <span
              style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
            >
              <span
                style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
              >
                <img
                  alt=""
                  aria-hidden="true"
                  src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%27100%27%20height=%27100%27/%3e"
                  style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
                />
              </span>
              <img
                alt="usdc"
                data-nimg="intrinsic"
                decoding="async"
                src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
              />
              <noscript />
            </span>
          </div>
          <div
            class="ml-2 flex flex-col justify-center text-sm"
          >
            <div
              class="font-bold"
            >
              USD Coin
            </div>
            <div
              class="uppercase"
            >
              usdc
            </div>
          </div>
        </div>
        <div
          class="text-right"
        >
          <div
            class="font-bold"
          >
            $ 
          </div>
          <div
            class="font-light text-sm text-gray-400"
          >
            0 usdc
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
